backend/tracim_backend/lib/utils/authentication.py

```diff
@@ -20,16 +28,195 @@
+        #     return Denied("Nobody is allowed")
+        return Allowed("Allowed")
+
+    def remember(
+        self, request: TracimRequest, userid: typing.Union[str, int], **kw: typing.Any
+    ) -> typing.Iterable[str]:
+        self._session_helper.remember(request, userid)
+        return []
+
+    def forget(self, request: TracimRequest, **kw: typing.Any) -> typing.Iterable[str]:
+        if "saml_user_id" in request.session:
+            del request.session["saml_user_id"]
+            del request.session["saml_mail"]
+            del request.session["saml_name"]
+        self._session_helper.forget(request, **kw)
+        return []
+
+    def _idp_chooser(self, request: TracimRequest) -> Response:
+        # TODO - SGD - 2023-07-26 - This view should render an HTML form
+        # with the list of configured IdPs.
+        # The form should redirect to the chosen IdPs SSO endpoint when submitted.
+        # Maybe some JS will be needed here?
+        return Response(
+            f"""
+          <form action="{self.app_config.WEBSITE__BASE_URL}/saml/acs" method="POST">
+            <input name="saml_user_id" type="hidden" value="saml-user"/>
+            <input name="redirect_url" type="hidden" value="{request.path_url}" />
+            <button>Login</button>
+          </form>"""
+        )
+
+    def _acs(self, request: TracimRequest) -> Response:
+        response = request.POST['SAMLResponse']
+        try:
+            authn_response = self.saml_client.parse_authn_request_response(
+                response,
+                binding=BINDING_HTTP_POST,
+            )
+        except ResponseLifetimeExceed as e:
+            return Response(e.__str__())
+        identity = authn_response.get_identity()
+        if identity is None or "UserID" not in identity:
+            return HTTPBadRequest()
+
+        # TODO - M.L - 2023/09/06 - Associate response / expiry date to session
+        request.session["saml_user_id"] = ''.join(identity["UserID"])
+        request.session["saml_mail"] = ''.join(identity["EmailAddress"])
+        request.session[
+            "saml_name"] = f'{"".join(identity["FirstName"])} {"".join(identity["LastName"])}'
+
+        return HTTPFound("/")
+
+    def _sso(self, request: TracimRequest) -> Response:
+        if "target" not in request.params:
+            return HTTPBadRequest("Missing target parameter")
+
+        target = None
+        for entity_id, data in self.saml_client.config.vorg.items():
+            if request.params["target"] == data.common_identifier:
+                target = entity_id
+
+        if target is None:
+            return HTTPNotFound("This IdP doesn't exist")
+        _, info = self.saml_client.prepare_for_authenticate(
+            entityid=target,
+        )
+
+        redirect_url = None
+        for key, value in info['headers']:
+            if key == 'Location':
+                redirect_url = value
+        return HTTPFound(redirect_url)
```

Comment:
This functions is the entrypoint for the SAML auth flow, it tries to get the IdP chosen name from the `target` query parameter and tries to match it with a `vorg` entry from the config (see details about the config below). If it finds a match, it redirects to it's sso endpoint.